### DISCUSSION TITLE
RfC - Headless django CMS via REST API - resolved ✅

### DISCUSSION THREAD

**Discussion by macolo** [django-cms/django-cms#7835] [General]
## Intro

- Please see the overall plan in the [Product Strategy 2023 document](https://docs.google.com/document/d/1wxKOMJaYCr9a0Tims_eOaG_hgtxrVCARnJmTxL4pIHY/edit)
- See the [Headless package for django-cms Brainstorm Session Notes 2023](https://docs.google.com/document/d/10_cxT11Rk4OqCux1AUJgajJfxtfyW_F-ThP3LSMOY3Y/)
- way of moving forward: iterative, step by step
- We are looking for sponsorship for this project

## Technical considerations
- much of django CMS editing interface is using django admin forms inside iframes. This is an approach that we can continue to make use of for headless implementations as well. 
- The big advantage of relying on django admin forms is that we can stand on the shoulders of "django admin" and the editor experience is seamless across django admin and django cms admin interfaces.
- We plan to achieve basic headless mode (v1) in the following way:
    - page tree and django admin via /admin url can be loaded in a separate browser tab - DONE
    - the page content editing sidebar and cms menu bar can be loaded in a separate browser tab - DONE
    - #7927 - DONE
    - a version of the website frontend can be loaded in a separate browser tab via versioning - DONE
    - the published version of the content models including page tree should be accessible read-only via REST API - DONE
- milestones:
   - make django CMS core ready for headless (see #7850)
   - implement the REST API for CMS page content and aliases (see https://github.com/fsbraun/djangocms-rest)

**Comment by jrief** [comment]
In case you are interested, I created a `RichtextArea` widget based on TipTap, where you can attach any arbitrary Django form. This all works with the new `<dialog>`-HTML element and does not require any Django-admin, nor does is need an `<iframe>`.

_Why is that important?_

Consider elements, such as `<a href="…">`, `<img src="…">`, `<button …>`, etc. where the user must enter more than one extra field.
For example a link: This could just be the URL of an external website, but often we need many more input fields, for instance if we want to link onto objects inside our database, so that they are independent of the URL mapping. Or if we want to specify the `target` or `rel` attributes.

When using images in richtext, it even becomes more complicate than that.  We then need a way to store the image's payload data somewhere on disk, whereas we want the metadata (width, height, alignment, etc.) inside the document. There is no one-fits-all solution for that, we want the customer to design their model describing the values they need.

In django-formset version 1.4 this will be possible. One then can create his own form definition in Python and attach it to a TipTap control element. Clicking on that button will open a dialog window with the fields declared in the mentioned form. This can be used for any kind of extra control elements inside your richtext editor. The data submitted by those dialog forms, then is merged
with the "rich" text and its payload can be stored in JSON in any database.

The latter has the advantage that no such thing as "text enabled plugins" are required anymore.

**Reply by macolo** [reply]
Just for visibility, I'll add to this that we are working on djangocms-text which will provide a satisfactory UX for text-enabled plugins. :)

**Reply by jrief** [reply]
Adopting django-formset would make sense, if django-CMS wants to avoid relying on the Django-Admin for editing plugins. Then plugins could be edited by just declaring a Django form. This also would prevent the need for an `<iframe>` serving Django's admin content. Of course, this would require a lot more effort to integrate.

**Comment by fsbraun** [comment]
We will also have to json-ify the data bridge: #7850

**Comment by ralfbraendli** [comment]
From my point of view, the most important part would be the possibility to add and edit plugins into placeholder via the Django Admin. This by it self enables everyone to use the Django CMS headless by implementing the API part on it's own. This could be a good first step into a  fully headlines CMS.

**Comment by macolo** [comment]
Quick update - in the meantime a lot of effort and progress has been made, please see djangocms-rest (https://github.com/fsbraun/djangocms-rest) and https://github.com/django-cms/django-cms/pull/7850 (Headless Readiness of the core) for further information.
